src/lib.rs

```diff
@@ -584,261 +596,276 @@ impl Github {
-                    )
+                    (parsed_url, Some(auth))
                 } else {
-                    debug!("App token is stale, refreshing");
-                    let token_ref = apptoken.access_key.clone();
-                    Box::new(
-                        self.app()
-                            .make_access_token(apptoken.installation_id)
-                            .and_then(move |token| {
-                                let auth = format!("token {}", &token.token);
-                                *token_ref.lock().unwrap() = Some(token.token);
-                                parsed_url
-                                    .map(|u| (u, Some(auth)))
-                                    .map_err(Error::from)
-                                    .into_future()
-                            }),
-                    )
+                    unreachable!("request() should be handling the refresh!");
                 }
             }
-            None => Box::new(
-                parsed_url
-                    .map(|u| (u, None))
-                    .map_err(Error::from)
-                    .into_future(),
-            ),
-        }
+            None => (parsed_url, None),
+        };
+
+        Ok(tuple)
     }
 
-    fn request<Out>(
+    async fn request<Out>(
         &self,
         method: Method,
         uri: &str,
         body: Option<Vec<u8>>,
         media_type: MediaType,
         authentication: AuthenticationConstraint,
-    ) -> Future<(Option<Link>, Out)>
+    ) -> Result<(Option<Link>, Out)>
     where
-        Out: DeserializeOwned + 'static + Send,
+        Out: DeserializeOwned + 'static + Send + Sync,
     {
-        let url_and_auth = self.url_and_auth(uri, authentication);
+        if let Some(&Credentials::InstallationToken(ref apptoken)) =
+            self.credentials(authentication)
+        {
+            if apptoken.token().is_none() {
+                debug!("App token is stale, refreshing");
+                let refresh_uri = self.host.clone()
+                    + &format!(
+                        "/app/installations/{}/access_tokens",
+                        apptoken.installation_id
+                    );
+                let url_and_auth = self
+                    .url_and_auth(&refresh_uri, AuthenticationConstraint::JWT)
+                    .await?;
+
+                let token_ref = apptoken.access_key.clone();
+
+                let token_response: app::AccessToken = self
+                    .authenticated_request(
+                        Method::POST,
+                        &refresh_uri,
+                        url_and_auth.0,
+                        url_and_auth.1,
+                        Some(Vec::new()),
+                        MediaType::Preview("machine-man"),
+                    )
+                    .await?
+                    .1;
 
+                *token_ref.lock().unwrap() = Some(token_response.token);
+            }
+        }
```

Comment:
I could not figure out how to make recursive async functions work, so I flattened it here. Ugly! :)

Reply:
that's when i stopped making hubcaps async/await ready :)

Reply:
When I started and knew almost nothing about async/await, I just deleted all the JWT code to avoid the problem ... and then added it back when I had a better feel for what was going on :)

Reply:
When I started and knew almost nothing about async/await, I just deleted all the JWT code to avoid the problem ... and then added it back when I had a better feel for what was going on :)

Reply:
When I started and knew almost nothing about async/await, I just deleted all the JWT code to avoid the problem ... and then added it back when I had a better feel for what was going on :)